### PR TITLE
Add documentation to public API

### PR DESCRIPTION
**Pull request by michaelabenyohai**
Also rename methods/variables to be more consistent.

### CODE REVIEW

**Comment by philandstuff** [comment]
needs rebase

**Comment by michaelabenyohai** [comment]
Rebased

File: src/test/java/uk/gov/verifiablelog/merkletree/MerkleTreeMockTests.java [hunk 1]
```diff
@@ -84,9 +84,9 @@ public void currentRoot_retrievesAndUsesMemoizationStoreHashes() {
     }
 
     @Test
-    public void pathToRootAtSnapshot_retrievesAndUsesMemoizationStoreHashes() {
+    public void auditProof_retrievesAndUsesMemoizationStoreHashes() {
         /**
-         *  Tree and memoization expectations for a consistency proof between trees of sizes 4 and 8
+         *  Tree and memoization expectations for a audit proof target node: 4, 8
```
Comment (philandstuff) [inline src/test/java/uk/gov/verifiablelog/merkletree/MerkleTreeMockTests.java]: should be `an audit`

File: src/main/java/uk/gov/verifiablelog/merkletree/MemoizationStore.java [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
 
+/**
+ * Defines the contract through which a consumer of a {@link MemoizationStore} can query and update a set of known
+ * Merkle Tree root hashes for intermediate subtrees of a particular size and start index.
+ */
 public interface MemoizationStore {
-    public void put(Integer start, Integer size, byte[] value);
-    public byte[] get(Integer start, Integer size);
+    /**
+     * Adds the root hash of a subtree to the set of known intermediate Merkle Tree root hashes.
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/MemoizationStore.java]: I think it should be clear that `put` does not need to always add the root hash of a subtree, and that `get` can always return `null` if it feels like it

File: src/main/java/uk/gov/verifiablelog/merkletree/MerkleTree.java [hunk 3]
```diff
@@ -2,45 +2,73 @@
...
 import java.security.MessageDigest;
 import java.util.*;
-import java.util.function.Function;
-import java.util.function.Supplier;
 
+/**
+ * An implementation of a Verifiable Log.
+ * @see <a href="https://tools.ietf.org/html/rfc6962">RFC 6962</a>
+ */
 public class MerkleTree {
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/MerkleTree.java]: we want to rename this at some point don't we?

**Comment by michaelabenyohai** [comment]
@philandstuff I've addressed your comments in the latest commit.

File: src/main/java/uk/gov/verifiablelog/dao/memoization/MemoizationStore.java [hunk 4]
```diff
@@ -15,10 +16,11 @@
...
+     * An implementation can return null for any start and size, even if an earlier {@link #put(Integer, Integer, byte[])}
+     * operation successfully added the subtree root hash to the known set.
      * @param start The zero-based index of the first leaf in the subtree
      * @param size The number of leaves in the subtree
      * @return The Merkle Tree root hash of the subtree if known, else null if not known
      */
     byte[] get(Integer start, Integer size);
-}
-
+}
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/dao/memoization/MemoizationStore.java]: you should set intellij to not do this

File: src/main/java/uk/gov/verifiablelog/merkletree/MerkleUtil.java [hunk 5]
```diff
@@ -6,15 +6,17 @@
 /**
  * A utility class that provides commonly used operations when working with Merkle Trees.
```
Comment (philandstuff) [inline src/main/java/uk/gov/verifiablelog/merkletree/MerkleUtil.java]: I see this class as _not_ part of the public interface of the library, so we can maybe document that anything may change at any time

**Pull request merged**